lib/mongo/server/connection_pool/available_stack.rb

```diff
@@ -300,31 +288,83 @@ def close_stale_sockets!
           end
 
           mutex.synchronize do
-            num_checked_out = pool_size - queue_size
+            num_checked_out = pool_size - stack_size
             min_size_delta = [(min_size - num_checked_out), 0].max
 
             to_refresh.each do |connection|
-              if queue.include?(connection)
+              if connections.include?(connection)
                 connection.disconnect!
-                if queue.index(connection) < min_size_delta
+                if connections.index(connection) < min_size_delta
                   begin; connection.connect!; rescue; end
                 end
               end
             end
           end
         ensure
-          check_count_invariants
+          check_count_invariants(false)
+        end
+
+        def connection_removed
+          @pool_size -= 1
+          if @pool_size < 0
+            # This should never happen
+            log_warn("ConnectionPool::AvailableStack: unexpected push")
+            @pool_size = 0
+          end
         end
 
         private
 
-        def dequeue_connection
-          deadline = Time.now + wait_timeout
+        def close_connection!(connection, reason)
+          connection_removed
+          connection.disconnect!
+
+          publish_cmap_event(
+            Monitoring::Event::ConnectionClosed.new(
+              reason,
+              @address,
+              connection.id,
+            )
+          )
+        end
+
+        def is_stale?(connection)
+          if connection.generation != @generation
+            close_connection!(connection, Monitoring::Event::ConnectionClosed::STALE)
+            true
+          end
+        end
+
+        def is_idle?(connection)
+          if connection && connection.last_checkin && max_idle_time
+            if Time.now - connection.last_checkin > max_idle_time
+              close_connection!(connection, Monitoring::Event::ConnectionClosed::IDLE)
+              true
+            end
+          end
+        end
+
+        def pop_connection(deadline)
+          mutex.synchronize do
+            get_connection(deadline)
+          end
+        end
+
+        def get_connection(deadline)
           loop do
-            return queue.shift unless queue.empty?
+            until connections.empty?
+              connection = connections.shift
+              return connection unless is_stale?(connection) || is_idle?(connection)
```

Comment:
If the connection is stale or idle and is not returned here, shouldn't it be closed?

Reply:
Done